### PR TITLE
HARP-7373: Point label persistence across zoom levels.

### PR DESCRIPTION
**Pull request by atomicsulfate**
Signed-off-by: Andres Mandado <andres.mandado-almajano@here.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/text/TextElementStateCache.ts [hunk 1]
```diff
@@ -82,24 +103,42 @@ export class TextElementStateCache {
      * @param disableFading `True` if fading is currently disabled, `false` otherwise.
      * @returns `True` if any textElementGroup was evicted from cache, false otherwise.
      */
-    update(time: number, clearVisited: boolean, disableFading: boolean) {
+    update(time: number, disableFading: boolean, findReplacements: boolean, zoomLevel: number) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementStateCache.ts]: please add the extra params to the documentation

File: @here/harp-mapview/lib/text/TextElementStateCache.ts [hunk 2]
```diff
@@ -194,4 +233,43 @@ export class TextElementStateCache {
...
+        }
+
+        const maxSqDistError = getDedupSqDistTolerance(zoomLevel);
+
+        // Other labels found with the same text. Check if they're near enough to be considered
+        // duplicates.
+        const entryCount = cachedEntries.length;
+        const elementPosition = element.points as THREE.Vector3;
+        let duplicateIndex: number = -1;
+        for (let i = 0; i < entryCount; ++i) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementStateCache.ts]: how do you prevent this checking against itself? as far as i can read, the TextElementState is part of this cache, so it is possible that the element is checked to see if it is a duplicate with itself?

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 3]
```diff
@@ -871,7 +872,7 @@ export class TextElementsRenderer {
                 ++this.m_initializedTextElementCount;
             }
         }
-        return textElement.loadingState === LoadingState.Initialized;
+        return textElement.glyphs !== undefined;
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: why has this changed?

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 4]
```diff
@@ -1687,7 +1692,12 @@ export class TextElementsRenderer {
             const textNeedsDraw = !textRejected || textRenderState!.isFading();
 
             if (textNeedsDraw) {
-                if (pointLabel.textReservesSpace) {
+                // Don't allocate space for rejected text. When zooming, this allows placement of a
+                // lower priority text element that was displaced by a higher priority one (not
+                // present in the new zoom level) before an even lower priority one takes the space.
+                // Otherwise the lowest priority text will fade in and back out.
+                // TODO: Add a unit test for this scenario.
+                if (pointLabel.textReservesSpace && !textRejected) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: is this related?

File: @here/harp-mapview/lib/text/TextElementStateCache.ts [hunk 5]
```diff
@@ -194,4 +233,43 @@ export class TextElementStateCache {
         this.m_referenceMap.set(textElementGroup, textElementGroupState);
         this.m_sortedGroupStates = undefined;
     }
+
+    private findDuplicate(
+        elementState: TextElementState,
+        zoomLevel: number
+    ): { entry: TextElementState[]; index: number } | undefined {
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementStateCache.ts]: should entry be "entries"

File: @here/harp-mapview/lib/text/TextElementStateCache.ts [hunk 6]
```diff
@@ -115,61 +154,61 @@ export class TextElementStateCache {
...
-
+            cacheResult.entry[cacheResult.index] = elementState;
             cachedDuplicate.reset();
             return true;
         }
 
         return false;
     }
 
+    replaceElement(zoomLevel: number, elementState: TextElementState): void {
```
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementStateCache.ts]: please add some documentation to this function.

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 5 above)
Comment (nzjony) [inline @here/harp-mapview/lib/text/TextElementStateCache.ts]: would it make sense to also return the duplicate? I notice there are a few cases where it writes: cacheResult.entry[cacheResult.index]; wouldn't it be a bit cleaner / simpler to just return this item?

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: done.

File: @here/harp-mapview/lib/text/TextElementStateCache.ts [hunk 7]
```diff
@@ -82,24 +103,42 @@ export class TextElementStateCache {
      * @param disableFading `True` if fading is currently disabled, `false` otherwise.
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: done

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: The duplicate cache is cleared after every frame and refilled again in a loop that iterates once through  every visited label. See the call to `deduplicateElement` in `prepareTextElementGroup`. So every visited label is added at most once per frame.

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 4 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: This is not related to your other comment above. This is a minor improvement that avoids flickering in particular scenario:
Imagine we are look at "EUROPE" label, then we zoom in to a level where continent labels are disabled and capitals become visible. "London" starts fading in to the west side (it doesn't collide with "EUROPE") as "EUROPE" keeps fading out. As soon as "EUROPE" becomes invisible, the label "DEUTSCHLAND" that was in collision both with "EUROPE" and "London" starts fading in, since it has less priority than "EUROPE" but more than "London", and causes the rejection of "London" that starts fading out.
If we let "DEUTSCHLAND" fade in as soon as "EUROPE" starts fading out, then "London" will never fade in in the first place.

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 6 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: done

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 3 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: When a label is replaced, the new label must inherit temporally the glyphs of the old label while its own glyphs are loaded asynchronously. Otherwise the replacement won't be rendered in the meantime. Nevertheless, `initializeGlyphs` will load the final glyphs for a label replacement and replace the inherited ones as soon as they're loaded.
So `initializeGlyphs` must return true as soon as a label has some glyphs assigned (either the final ones or the temporally inherited).

Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: I'll document this in the code.

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 5 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: returning only  cacheResult.entry[cacheResult.index] won't be enough cause sometimes the caller needs to overwrite that particular entry, i.e, it needs to do:
cacheResult.entry[cacheResult.index] = newLabel;

We could return the three things: the entries array, the index and the content of the entry addressed by the index, then the object we use for the result would look like this:
`const tmpCachedDuplicate: { entries: TextElementState[]; index: number; entry:TextElementState|undefined } = {
    entries: [],
    index: -1,
    entry:undefined
};`
But then it's confusing because there's more than one way to get the same stuff, you can get `entry` also by doing `entries[index]`. Also, someone might feel tempted to do `entry = newElementState` to replace the content of an entry, which wouldn't work. And we can't even set `entry` to `readonly` cause the object is reused on every call.

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 3 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: There some comment in `TextElement.replace()`, but I'll add more in `initializeGlyphs`

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 5 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: done

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 4 above)
Comment (nzjony) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: Ok, please move to a separate PR, if there is any regression, it would be easier to revert this single line than have to revert everything.

File: @here/harp-mapview/lib/text/TextElementStateCache.ts (hunk 5 above)
Comment (nzjony) [reply @here/harp-mapview/lib/text/TextElementStateCache.ts]: thanks for taking a look, I understand.

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts (hunk 4 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: I've moved it to a separate commit, so that it could reverted if necessary as you say.
But I left it in the same PR, i won't squash it when merging. The change is related to zoom level persistence since the scenario I described is reproducible as soon as zoom level persistence is introduced.

Comment (nzjony) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: good idea, yes it can be a separate commit, cool!

Comment (nzjony) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: wait, separate commit is cool, but please merge the other two commits.

Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: will do
